huaweicloud/services/waf/resource_huaweicloud_waf_cloud_instance.go

```diff
@@ -0,0 +1,481 @@
+		EnterpriseProjectId:        common.GetEnterpriseProjectID(d, cfg),
+	}
+	orderId, err := clouds.Create(wafClient, opts)
+	if err != nil {
+		return diag.Errorf("error creating cloud WAF: %s", err)
+	}
+
+	bssClient, err := cfg.BssV2Client(region)
+	if err != nil {
+		return diag.Errorf("error creating BSS v2 client: %s", err)
+	}
+	err = common.WaitOrderComplete(ctx, bssClient, *orderId, d.Timeout(schema.TimeoutCreate))
+	if err != nil {
+		return diag.Errorf("[DEBUG] The order is not completed while creating cloud WAF: %#v", err)
+	}
+	resourceId, err := common.WaitOrderResourceComplete(ctx, bssClient, *orderId, d.Timeout(schema.TimeoutCreate))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(resourceId)
+
+	return resourceCloudInstanceRead(ctx, d, meta)
+}
+
+func analysisSpecCode(specCode string) (string, error) {
+	// All specification codes:
+	// + waf.detection
+	// + waf.instance.professional
+	// + waf.enterprise
+	// + waf.ultimate
+	re := regexp.MustCompile(`^waf\.(?:[a-z]+\.)?(\w+)$`)
+	result := re.FindStringSubmatch(specCode)
+	// The right length of the string match is two, first is the match result of the regex string,
+	// second is the match result of the regex group.
+	if len(result) < 2 {
+		return "", fmt.Errorf("invalid specification code, want 'waf.xxx' or 'waf.xxx.xxx', but '%s'.", specCode)
+	}
+	return result[1], nil
+}
+
+func QueryCloudInstance(client *golangsdk.ServiceClient, instanceId string) (*clouds.Instance, error) {
+	default404Err := golangsdk.ErrDefault404{
+		ErrUnexpectedResponseCode: golangsdk.ErrUnexpectedResponseCode{
+			Body: []byte(fmt.Sprintf("the cloud WAF (%s) does not exist", instanceId)),
+		},
+	}
+	resp, err := clouds.Get(client)
+	if err != nil {
+		return nil, err
+	}
+	if resp == nil || len(resp.Resources) < 1 {
+		return nil, default404Err
+	}
+	for _, val := range resp.Resources {
+		if val.Type == string(ResourceTypeInstance) && val.ID == instanceId {
+			return resp, nil
+		}
+	}
+	return nil, default404Err
+}
+
+func flattenExtendedPackages(res clouds.Resource) []map[string]interface{} {
+	return []map[string]interface{}{
+		{
+			"resource_size": res.Size,
+		},
+	}
+}
+
+func resourceCloudInstanceRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	wafClient, err := cfg.WafV1Client(region)
+	if err != nil {
+		return diag.Errorf("error creating WAF v1 client: %s", err)
+	}
+
+	instanceId := d.Id()
+	resp, err := QueryCloudInstance(wafClient, instanceId)
+	if err != nil {
+		return common.CheckDeletedDiag(d, err, "cloud WAF")
+	}
+	log.Printf("[DEBUG] The resources list of the cloud WAF is: %#v", resp.Resources)
+
+	mErr := multierror.Append(nil,
+		d.Set("region", cfg.GetRegion(d)),
+	)
+	for _, val := range resp.Resources {
+		if val.Type == string(ResourceTypeInstance) {
+			specCode, err := analysisSpecCode(val.SpecCode)
```

Comment:
the spec_code comes form the API response, why check it here?

Reply:
because of the request argument and the response attribute are different.
The value of the request argument is `detection`, but the response is `waf.detection`.

Reply:
Some comments have been added.